In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("../data/online_retail.csv")

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2022-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2022-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.shape

(541909, 8)

In [6]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

In [7]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 541909
Columns: 8


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


In [9]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [10]:
df = df.dropna(subset=['CustomerID'])

In [11]:
df.shape

(406829, 8)

In [12]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [13]:
df.duplicated().sum()

np.int64(5225)

In [14]:
df = df.drop_duplicates()

In [15]:
df.duplicated().sum()

np.int64(0)

In [16]:
df[df["InvoiceNo"].astype(str).str.startswith("C")].shape

(8872, 8)

In [17]:
df[df["InvoiceNo"].astype(str).str.startswith("C")].head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2022-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2022-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2022-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2022-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2022-12-01 10:24:00,0.29,17548.0,United Kingdom


In [18]:
df = df[
    ~df["InvoiceNo"].astype(str).str.startswith("C")
]

In [19]:
df[df["InvoiceNo"].astype(str).str.startswith("C")].shape

(0, 8)

In [20]:
df[df["Quantity"] <= 0].shape

(0, 8)

In [21]:
df[df["Quantity"] <= 0].head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


## Quantity Validation

- After removing cancelled invoices, no records with Quantity ≤ 0 were found.
- Therefore, no additional quantity cleaning was required.

In [22]:
df[df["UnitPrice"] <= 0].shape

(40, 8)

In [23]:
df[df["UnitPrice"] <= 0].head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2022-12-05 14:02:00,0.0,12647.0,Germany
33576,539263,22580,ADVENT CALENDAR GINGHAM SACK,4,2022-12-16 14:36:00,0.0,16560.0,United Kingdom
40089,539722,22423,REGENCY CAKESTAND 3 TIER,10,2022-12-21 13:45:00,0.0,14911.0,EIRE
47068,540372,22090,PAPER BUNTING RETROSPOT,24,2023-01-06 16:41:00,0.0,13081.0,United Kingdom
47070,540372,22553,PLASTERS IN TIN SKULLS,24,2023-01-06 16:41:00,0.0,13081.0,United Kingdom


In [25]:
df = df[df["UnitPrice"] > 0]

In [26]:
df[df["UnitPrice"] <= 0].shape

(0, 8)

## Unit Price Validation

- 40 records contained UnitPrice ≤ 0.
- These records do not represent valid sales transactions.
- Such records can affect revenue calculations and customer monetary analysis.
- Therefore, all records with UnitPrice ≤ 0 were removed.

In [27]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [29]:
df["Revenue"] = df["Quantity"] * df["UnitPrice"]

## Revenue Feature Engineering

A new column named Revenue was created using:

Revenue = Quantity × UnitPrice

This column represents the total sales amount generated by each transaction and will be used for:

- Revenue Analysis
- Customer Monetary Value Calculation
- RFM Analysis
- Customer Segmentation
- Business Insights

In [30]:
df.info()

<class 'pandas.DataFrame'>
Index: 392692 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392692 non-null  str           
 1   StockCode    392692 non-null  str           
 2   Description  392692 non-null  str           
 3   Quantity     392692 non-null  int64         
 4   InvoiceDate  392692 non-null  datetime64[us]
 5   UnitPrice    392692 non-null  float64       
 6   CustomerID   392692 non-null  float64       
 7   Country      392692 non-null  str           
 8   Revenue      392692 non-null  float64       
dtypes: datetime64[us](1), float64(3), int64(1), str(4)
memory usage: 30.0 MB


In [31]:
df.to_csv("../data/cleaned_online_retail.csv", index=False)

# Data Cleaning Summary

Initial Dataset Size:
- Rows: 541,909
- Columns: 8

Data Cleaning Steps Performed:
1. Removed rows with missing CustomerID.
2. Removed duplicate records.
3. Removed cancelled invoices (InvoiceNo starting with 'C').
4. Verified no invalid quantities remained.
5. Removed records with UnitPrice ≤ 0.
6. Converted InvoiceDate to datetime format.
7. Created Revenue column using Quantity × UnitPrice.

Final Dataset Size:
- Rows: 392,692
- Columns: 9

The cleaned dataset is ready for Exploratory Data Analysis (EDA), RFM Analysis, Customer Segmentation, and Product Recommendation.